### PR TITLE
[NO-JIRA] Add arch doc

### PR DESCRIPTION
**Pull request by tiagoengel**
+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/master/CONTRIBUTING.md)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
+## Backpack-android architecture overview:
+
+Backpack-android is a Android a traditional gradle Android lib project, composed of two modules.
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
`backpack-android is a traditional Gradle Android lib project, composed of two modules.
```

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+- app: The example app
+- Backpack: The Backpack Android components
+
+## App
+
+### Stories
+
+Stories are defined programmatically in tree like structure that is then used to build the UI.
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
Stories are defined programmatically in  tree-like structure that is then used to build the UI.
```

File: docs/ARCHITECTURE.md (hunk 1 above)
Comment (georgegillams) [inline docs/ARCHITECTURE.md]:
```suggestion
Backpack-android is a traditional gradle Android lib project, composed of two modules.
```

File: docs/ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
+## Backpack-android architecture overview:
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
## backpack-android architecture overview:
```

File: docs/ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  "Panel" story NodeData { Story of R.layout.fragment_panel },
+  "Spinner" story NodeData({ children -> SubStory of children },
+    mapOf(
+      "Default" story NodeData { Story of R.layout.fragment_spinner },
+      "Small" story NodeData { Story of R.layout.fragment_spinner_small }
+    )),
+  ```
+
+- **`res/layout/`**
+  Where layouts for each story is defined. They are named `fragment_{story_name}`
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
  Where layouts for each story is defined. They are named `fragment_{story_name}`.
```

File: docs/ARCHITECTURE.md (hunk 2 above)
Comment (georgegillams) [inline docs/ARCHITECTURE.md]:
```suggestion
Stories are defined programmatically in dendroidal structure that is used to build the UI.
```

File: docs/ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+- **`{javaSrc}/BackpackDemoApplication`**
+  Where all initialization logic resides, including registering the current theme.
+
+- **`{javaSrc}/data/SharedPreferences`**
+  Used to read/write shared preferences for the example app, including the current theme.
+
+
+## Backpack
+
+The backpack lib is where all components live. All packages, except `utils` are Backpack components.
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
The `backpack` lib is where all components live. All packages, except `utils` are Backpack components.
```

File: docs/ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+### Component's architecture
+
+There are only a few things that should be observed about the architecture of the components: 
+
+1. Whenever possible the Backpack component should extend from the Android base component and add the additional extra functionality instead of wrapping the Android component inside another `ViewGroup`. This is to keep the view hierarchy as small as possible.
+2. All props should be available via XML and programmatically. Whenever a component is available in both XML and code all its properties should also be available in both. 
+3. Changing a property programmatically should immediately reflect on the UI.
+4. Components should be "extensible", i.e. `open class` in Kotlin. It's also important to think about the component being used that way when writing/reviewing it.
+
+More can be found here: https://github.com/Skyscanner/backpack-android/tree/master/decisions
```
Comment (georgegillams) [inline docs/ARCHITECTURE.md]: Maybe this should live in it's own `### decisions` section?

File: docs/ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+### Theming
+- Uses default Android theming mechanism https://developer.android.com/guide/topics/ui/look-and-feel/themes
```
Comment (georgegillams) [inline docs/ARCHITECTURE.md]:
```suggestion
- Uses the native Android theming mechanism https://developer.android.com/guide/topics/ui/look-and-feel/themes
```

File: docs/ARCHITECTURE.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+### Caveats
+
+The `style` attribute will take precedence over the global `bpkChipStyle`, meaning it is possible to
+have a different style for a single component.
+
+It is also possible to set theming properties directly via `xml`:
+```xml
+  <BpkChip style="@style/RedChip" app:chipSelectedBackgroundColor="@colors/bpkRed500" />
+```
+This is not advertised and is not how we intend theming to be used.
```
Comment (georgegillams) [inline docs/ARCHITECTURE.md]:
```suggestion
This is not advertised and is not a supported way for theming to be used. If you do this, we cannot guarantee that small changes to Backpack will not break your UI.
```

File: docs/ARCHITECTURE.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  - Live inside the `Backpack` module and run inside an emulator, hence "connected". This includes mostly unit testing for visual components.
+
+- Unit tests:
+  - Live inside the `Backpack` module and do not need an emulator to run. This includes all other unit tests that don't require an emulator.
+
+- Snapshot tests:
+  - Live inside the `app` module and run inside an emulator. Some of those tests are very flaky in the 
+  CI and will be annotated with `@FlakyTest`. This means they will not run on CI but still run locally before a release.
+
+
```
Comment (georgegillams) [inline docs/ARCHITECTURE.md]: Such line break. Many wow.

File: docs/ARCHITECTURE.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+```
+This is not advertised and is not how we intend theming to be used.
+
+
+## Testing
+
+There are currently three types of tests in `backpack-android`:
+
+- Connected tests:
+  - Live inside the `Backpack` module and run inside an emulator, hence "connected". This includes mostly unit testing for visual components.
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
  - Live inside the `backpack` module and run inside an emulator, hence "connected". This includes mostly unit testing for visual components.
```

Not sure which of capitalised or lower case is better for the module name but it's done in two ways in this file, so my suggestions change it to lower case everywhere. If that isn't right though go for capitalised 😄

File: docs/ARCHITECTURE.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+- Unit tests:
+  - Live inside the `Backpack` module and do not need an emulator to run. This includes all other unit tests that don't require an emulator.
```
Comment (shaundon) [inline docs/ARCHITECTURE.md]:
```suggestion
  - Live inside the `backpack` module and do not need an emulator to run. This includes all other unit tests that don't require an emulator.
```

File: docs/ARCHITECTURE.md (hunk 2 above)
Comment (tiagoengel) [reply docs/ARCHITECTURE.md]: I have no idea what `dendroidal` means lol. I'll use @shaundon suggestion instead as I think is easier to understand.

File: docs/ARCHITECTURE.md (hunk 10 above)
Comment (tiagoengel) [reply docs/ARCHITECTURE.md]: well the actual gradle module name is upper case. I'll review it and make them upper case

File: docs/ARCHITECTURE.md (hunk 6 above)
Comment (tiagoengel) [reply docs/ARCHITECTURE.md]: That's a good point. Can we leave it here for now and I can move it do decision files in the future?

**Pull request merged**